src/rgw/driver/d4n/d4n_policy.cc

```diff
@@ -395,6 +409,175 @@ bool LFUDAPolicy::erase(const DoutPrefixProvider* dpp, const std::string& key, o
+    	ldpp_dout(dpp, 10) << __func__ << " : "  << __LINE__ << ": bucket name is " << c_bucket->get_name() << dendl;
+	/*
+    	c_bucketinfo = bucket->get_info();
+    	ldpp_dout(dpp, 10) << __func__ << " : "  << __LINE__ << ": bucket name is " << b_name << dendl;
+	c_bucket = driver->get_bucket(c_bucketinfo);
+    	ldpp_dout(dpp, 10) << __func__ << " : "  << __LINE__ << ": bucket name is " << c_bucket->get_name() << dendl;
+	*/
+	/*
+ 	int ret = driver->get_bucket(dpp, nullptr, c_rgw_bucket, &c_bucket, null_yield);
+	if (ret < 0){
+    		ldpp_dout(dpp, 10) << __func__ << " : "  << __LINE__ << " cleaning get_bucket() failed for Bucket: " << b_name << dendl;
+		continue;
+	}
+	*/
+
+	std::unique_ptr<rgw::sal::Object> c_obj = c_bucket->get_object(c_obj_key);
+
+	std::unique_ptr<rgw::sal::Writer> processor =  driver->get_atomic_writer(dpp,
+				  null_yield,
+				  c_obj.get(),
+				  c_user->get_id(),
+				  NULL,
+				  0,
+				  "");
+
+  	int op_ret = processor->prepare(null_yield);
+  	if (op_ret < 0) {
+    	  ldpp_dout(dpp, 20) << "processor->prepare() returned ret=" << op_ret << dendl;
+    	  break;
+  	}
+
+	std::string prefix = b_name+"_"+key;
+	off_t lst = it->second->size;
+  	off_t fst = 0;
+  	off_t ofs = 0;
+
+	
+  	rgw::sal::DataProcessor *filter = processor.get();
+	do {
+    	  ceph::bufferlist data;
+    	  if (fst >= lst){
+      	    break;
+    	  }
+    	  off_t cur_lst = std::min<off_t>(fst + cct->_conf->rgw_max_chunk_size, lst);
+    	  std::string oid_in_cache = "D_" + prefix + "_" + std::to_string(fst) + "_" + std::to_string(cur_lst);  	  
+    	  std::string new_oid_in_cache = prefix + "_" + std::to_string(fst) + "_" + std::to_string(cur_lst);
+      	  ldpp_dout(dpp, 20) << __func__  << ": " << __LINE__ << ": oid_in_cache is: " << oid_in_cache << dendl;
+    	  cacheDriver->get(dpp, oid_in_cache, fst, cur_lst, data, obj_attrs, null_yield);
+    	  len = data.length();
+    	  fst += len;
+
+    	  if (len == 0) {
+      	    break;
+   	  }
+
+      	  ldpp_dout(dpp, 20) << __func__  << ": " << __LINE__ << ": data is: " << data.to_str() << dendl;
+    	  op_ret = filter->process(std::move(data), ofs);
+    	  if (op_ret < 0) {
+      	    ldpp_dout(dpp, 20) << "processor->process() returned ret="
+          	<< op_ret << dendl;
+      	    return;
+    	  }
+
+          ldpp_dout(dpp, 20) << "AMIN: " << __func__ << __LINE__ << dendl;
+  	  rgw::d4n::CacheBlock block;
+    	  block.cacheObj.bucketName = c_obj->get_bucket()->get_name();
+    	  block.cacheObj.objName = c_obj->get_key().get_oid();
+      	  block.size = len;
+     	  block.blockID = ofs;
+          ldpp_dout(dpp, 20) << "AMIN: " << __func__ << __LINE__ << dendl;
+	  op_ret = dir->update_field(&block, "dirty", "false", null_yield); 
+    	  if (op_ret < 0) {
+      	    ldpp_dout(dpp, 20) << "updating dirty flag in Block directory failed!" << dendl;
+      	    return;
+    	  }
+
+    	  cacheDriver->rename(dpp, oid_in_cache, new_oid_in_cache, null_yield);
```

Comment:
are you renaming the dirty blocks to remove the prefix 'D' once they are written to the storage backend? Wasn't it decided that blocks/objects would either exist in the cache or in the storage backend? and hence delete them from the cache backend, once they are written back to the storage backend.

Reply:
no, the delete action should be done through LFUDA policy. There is a high probability that the written data will be read soon. an example is ML workloads.